### ISSUE TITLE
consider removing no-did-mount-set-state from recommended rules

### ISSUE THREAD

**Issue by jedwards1211** [yannickcr/eslint-plugin-react#596]
Yes it leads to layout thrashing, but what else can you do when you need to perform a JS-computed layout?  One can work around this by using DOM measuring wrapper components, but for components that do JS layout, setState() in componentDidMount() is a fairly common way of doing it.

**Comment by yannickcr** [comment]
I am agree with you, actually I'm doing exactly the same thing in some of my projects (render, measure, setState, re-render).

But since it is pretty easy to do some accidental layout trashing without even realizing it I think it is better to keep this rule in the recommended ones, and use a `/* eslint-disable react/no-did-mount-set-state */` comment when justified.

**Comment by mjackson** [comment]
It's absolutely necessary to do a `setState` inside `componentDidMount` for doing things like measuring. Since there's no other way to do this, I'd agree with @jedwards1211 that it should be removed from the "recommended" rules.

@yannickcr Maybe it's time for a `react/warnings` config? [eslint-plugin-import does this](https://github.com/benmosher/eslint-plugin-import/blob/9fb72740060fa7c722f5eb8546b3a9cfc93a19b2/src/index.js#L29-L30) and it's great. You simply extend `import/warnings` if you want to see these kinds of opinionated error messages and `import/errors` if you only want to see actual error cases. Could be a solution to #636 as well.

**Comment by jedwards1211** [comment]
Well one *could* dispatch Redux actions and it will be interesting to see if anyone comes up with a really slick fully-JS layout solution for React components, but when one just needs a few workarounds for the limitations of CSS layout, then component `state` is a really logical place to put the layout info.

**Comment by jedwards1211** [comment]
Of course that begs the question: eslint leaders wouldn't consider dispatching Redux actions in `componentDidMount` a code smell, I hope?

**Comment by mjackson** [comment]
"Despatch a Redux action" is just a really elaborate way of setting state.

**Comment by mjackson** [comment]
@yannickcr If you think `react/warnings` and `react/errors` might be a good approach, I can make a PR with what I think would be appropriate. Interested?

**Comment by jedwards1211** [comment]
@mjackson my point exactly

**Comment by yannickcr** [comment]
@mjackson that's an interesting approach, even if I mostly agree with @ljharb about warnings https://github.com/yannickcr/eslint-plugin-react/pull/636#issuecomment-226657532.

I'll be definitely interested if you have the time to craft a PR with this.

**Comment by SirCmpwn** [comment]
Another point: async componentDidMount functions.

**Comment by yannickcr** [comment]
@SirCmpwn I'm not sure about what you are referring to. Should not the [`allow-in-func`](https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/no-did-mount-set-state.md#allow-in-func-mode) option cover this case?

**Comment by SirCmpwn** [comment]
That option seems to be designed for this case, but it doesn't work with async/await keywords from ES7. Example:

```js
async componentDidMount() {
    const thing = await foobar();
    this.setState({ thing });
}
```

**Comment by ljharb** [comment]
(nit: async/await is not in ES7, it's stage 3 and just a "future ES proposal")

**Comment by SirCmpwn** [comment]
Nit sufficiently picked, still an issue :)

**Comment by yannickcr** [comment]
Since there is some legitimate use cases to use `setState` here I have removed these rules from the recommended configuration. The rules are still available for who want to activate them manually.
